### DISCUSSION TITLE
Mutable and hash

### DISCUSSION THREAD

**Discussion by cgabard** [sqlalchemy/sqlalchemy#6018] [Usage Questions]
Hello,

First thank you all for your wonderfull job on the project.

I have a question regarding about [mutation tracking](https://docs.sqlalchemy.org/en/13/orm/extensions/mutable.html)

Basically, in my use case, I want to use a Mutable on a dataclass mapped to a table. Something like this:

```python
class MyEnum(Enum):
    ...

my_table = Table(
    "my_tables",
    metadata,
    Column("slug", String(255), primary_key=True),
    Column("values", MutableList.as_mutable(ARRAY(Enum(MyEnum)), nullable=False),
)

@dataclass
class MyModel:
    slug: str
    values: List[MyEnum]

mapper_registry = registry()
mapper_registry.map_imperatively(MyModel, my_table)
```

But with that, creating instance of  `MyModel` is not possible because of the mutation tracking. It seems that the mutable parent need to be hashable.

I can add an `__hash__` method but it is a bit strange: hash should only be implemented for immutable types where, by definition, if I want to add a `MutableList`, I will update it. So it is a bit strange because to be able to use mutation tracking, I need to implement a "for immutable only" method.

Is there is something that I missed? Or does there is something that I can do to avoid that/other solutons? Any suggestions?
Thanks!

**Comment by zzzeek** [comment]
> Hello,
> 
> First thank you all for your wonderfull job on the project.

thanks!

> 
> But with that, creating instance of `MyModel` is not possible because of the mutation tracking. It seems that the mutable parent need to be hashable.

just so you know, the last time I looked at how "mutable" worked was years ago, so I didn't recall why this would be (though it sounds like something I would do).   The dataclasses thing is brand new and I don't know too much about them either besides what I had to do to get them working (they arent hashable?  that's weird) So I had to extraploate your fragment into a program so I could understand what we were talking about:


```python

from dataclasses import dataclass
import enum
from typing import List

from sqlalchemy import Column
from sqlalchemy import MetaData
from sqlalchemy import String
from sqlalchemy import Table
from sqlalchemy.ext.mutable import MutableList
from sqlalchemy.orm import registry
from sqlalchemy.sql.sqltypes import ARRAY
from sqlalchemy.sql.sqltypes import Enum


metadata = MetaData()


class MyEnum(enum.Enum):
    RED = 1
    GREEN = 2
    BLUE = 3


my_table = Table(
    "my_tables",
    metadata,
    Column("slug", String(255), primary_key=True),
    Column(
        "values", MutableList.as_mutable(ARRAY(Enum(MyEnum))), nullable=False
    ),
)


@dataclass
class MyModel:
    slug: str
    values: List[MyEnum]

    # this is the part needed for it to work
    def __hash__(self):
        return id(self)


mapper_registry = registry()
mapper_registry.map_imperatively(MyModel, my_table)

m1 = MyModel(values=[MyEnum.RED, MyEnum.GREEN], slug="slug")
```

so yes it seems like if `__hash__` is not there we get a stack trace as follows:
```

 File "/home/classic/dev/sqlalchemy/lib/sqlalchemy/ext/mutable.py", line 485, in set_
    value._parents[target.obj()] = key
  File "/opt/python-3.8.3/lib/python3.8/weakref.py", line 396, in __setitem__
    self.data[ref(key, self._remove)] = value
TypeError: unhashable type: 'MyModel'
```

this is because the Mutable._parents collection needs to track all ORM mapped instances that have a value for this specific key.   it's not any different from the Session's identity map except in terms of implementation it's placing the object in a WeakKeyDictionary, so that we can find it later.

> 
> I can add an `__hash__` method but it is a bit strange: hash should only be implemented for immutable types where, by definition, if I want to add a `MutableList`, I will update it. So it is a bit strange because to be able to use mutation tracking, I need to implement a "for immutable only" method.

that's the notion of `__hash__` from a data perspective.   The notion of `__hash__` from the perspective of the object management system is looking for the identity of your object, and that is immutable - id(object) will always return the same value.

this is an apparent limitation in the mutable extension that it is requiring that `__hash__` work in this way, where it must semantically resolve to the identity of the object, and that is not an assumption made by the ORM, which uses a more indirect approach of tracking objects.   it's not immediately apparent why the mutation tracking extension cannot use this approach as well.    So I have created and implemented #6020 , we will include that in 1.4 and we will see anything comes up then.

**Reply by cgabard** [reply]
First, sorry for not providing all details. Will do my best to provide a complete working example next time.

Your fragment, and error, are exactly my error.

> The dataclasses thing is brand new and I don't know too much about them either besides what I had to do to get them working (they arent hashable? that's weird)
> [...]
> that's the notion of __hash__ from a data perspective. The notion of __hash__ from the perspective of the object management system is looking for the identity of your object, and that is immutable - id(object) will always return the same value.

According to [dataclass documentation](https://docs.python.org/3/library/dataclasses.html), it do not generate `__hash__` for this reason. I can not link to the section but search for `unsafe_hash` parameters. Documentation explains that dataclass, by default, only generates hash if the class is set as `frozen` (immutable them). There is this paragraph particularly:

> `__hash__()` is used by built-in `hash()`, and when objects are added to hashed collections such as dictionaries and sets. Having a `__hash__()` implies that instances of the class are immutable. Mutability is a complicated property that depends on the programmer’s intent, the existence and behavior of `__eq__()`, and the values of the eq and frozen flags in the `dataclass()` decorator.
>
> By default, `dataclass()` will not implicitly add a `__hash__()` method unless it is safe to do so. Neither will it add or change an existing explicitly defined `__hash__()` method. [...]

So I guess it is a "purity" choice. I also tried adding the `unsafe_hash` parameter to dataclass but does not work neither because `MutableList` needs to be hashable too.

> that's the notion of `__hash__` from a data perspective. The notion of `__hash__` from the perspective of the object management system is looking for the identity of your object, and that is immutable - `id(object)` will always return the same value.

I get it but opened this discussion because I was not sure about it. If it was the attended behavior, I guess the mutable code can use `id(object)`. So I made the assumption that I missed something. It was my first time checking the Mutable and sqlalchemy code, so I am not very confident about my knowledge of it.

> that is not an assumption made by the ORM, which uses a more indirect approach of tracking objects

For the sake of curiosity, did you have any entry point to allow me understand how the ORM do it? Somewhere in the doc or source code ?

> So I have created and implemented #6020 , we will include that in 1.4 and we will see anything comes up then.

Thank you very much, I did not expect a so quick answer.

**Reply by zzzeek** [reply]
I understand the dataclass thing and absolutely, a mutable object shouldn't be hashable, SQLAlchemy has already considered this years ago when we made sure the ORM never consulted `__hash__` on an object, which was quite a big deal.   So I'm not sure why I so casually made use of `__hash__` when I did the mutable extension.   Anyway, the patch I've proposed passes tests (the reason I do them so quickly and immediately is because if I don't, then I'll never get to it, and this is a major flaw) so I've merged and resolved #6020.  the dataclasses feature is why this has come to the forefront, thanks for bringing my attention to it.

**Comment by dbarnett** [comment]
FYI, the "mutation tracking" link was broken for me and looks like it should point to https://docs.sqlalchemy.org/en/14/orm/extensions/mutable.html.
